### PR TITLE
Add serialization of typenode for null/undefined/never as part of metadata type

### LINKED ISSUE #12684: Verbose Debug Information: Node 94 (NullKeyword) was unexpected on typescript 2.1.1
**TypeScript Version:**  2.1.1 / nightly (2.2.0-dev.20161206)

When I compile my angular2 project with webpack and awesome-typescript-loader, get errors: 

```ts
[at-loader] Using typescript@2.2.0-dev.20161206 from typescript and "tsconfig.json" from /Users/zhouweiming/data/project/iermu_opssite/src/tsconfig.json

/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:62645
                throw e;
                ^

Error: Debug Failure. False expression: Unexpected node.
Verbose Debug Information: Node 94 (NullKeyword) was unexpected.
    at Object.assert (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:3177:23)
    at Object.Debug.failBadSyntaxKind (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:44765:55)
    at serializeTypeNode (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46715:30)
    at serializeTypeNode (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46683:56)
    at serializeTypeOfNode (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46544:28)
    at serializeParameterTypesOfNode (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46596:42)
    at addOldTypeMetadata (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46471:98)
    at addTypeMetadata (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46462:17)
    at transformAllDecoratorsOfDeclaration (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46301:13)
    at generateConstructorDecorationExpression (/Users/zhouweiming/data/project/iermu_opssite/src/node_modules/.typescript@2.2.0-dev.20161206/lib/typescript.js:46411:40)

```

get same error on typescript@2.1.1 and typescript@2.0.10 and typescript@2.2.0-dev.20161206.

`other env`:
+ mac os: 10.12.1
+ node: 7.2.0
+ webpack: 1.13.3
+ angular: 2.2.4
+ awesome-typescript-loader: 3.0.0-beta.9

When I remove `@Injectable()` from all my codes, my project can compile success, but in browser I get errors: can't resolve all params in a class who depend another `@Injectable()` class.

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #12684 and #11933

### CODE REVIEW

File: src/compiler/transformers/ts.ts [hunk 1]
```diff
@@ -1715,27 +1717,35 @@ namespace ts {
                 case SyntaxKind.UnionType:
                     {
                         const unionOrIntersection = <UnionOrIntersectionTypeNode>node;
-                        let serializedUnion: Identifier;
+                        let serializedUnion: Identifier | VoidExpression;
```
Comment (mhegazy) [inline src/compiler/transformers/ts.ts]: We should make the serializeType* family strongly typed when it comes to return value. today it returns `Expression`

File: src/compiler/transformers/ts.ts [hunk 2]
```diff
@@ -1715,27 +1717,35 @@ namespace ts {
...
+                                if (serializedUnion && isIdentifier(serializedUnion) && serializedUnion.text !== serializedIndividual.text) {
+                                    serializedUnion = undefined;
+                                    break;
+                                }
 
-                            // Different types
-                            if (serializedUnion && serializedUnion.text !== serializedIndividual.text) {
+                                serializedUnion = serializedIndividual;
+                            }
+                            else if (isVoidExpression(serializedIndividual)) {
```
Comment (mhegazy) [inline src/compiler/transformers/ts.ts]: it could also return `BinaryExpression` would that be fine?

Comment (mhegazy) [inline src/compiler/transformers/ts.ts]: or `ConditionalExpression`

**Comment by sheetalkamat** [comment]
@mhegazy Please take a look, I have handled "non identifier" serialized expressions with null/undefined/never in the update.

**Comment by mhegazy** [comment]
@sheetalkamat does this handel https://github.com/Microsoft/TypeScript/issues/13098 as well?

File: src/compiler/transformers/ts.ts [hunk 3]
```diff
@@ -1815,14 +1826,15 @@ namespace ts {
             }
         }
 
+        type SerializedEntityNameAsExpression = Identifier | BinaryExpression | PropertyAccessExpression;
```
Comment (mhegazy) [inline src/compiler/transformers/ts.ts]: nit, can you move this next to the other type alias. possibly move both to the top of the file.

**Comment by sheetalkamat** [comment]
@mhegazy #13098 is different. This only deals with union/intersection. Will take a look at that one separately.

**Pull request merged**